File: rbm-rfm-frontend/src/types/projects.ts
/**
 * Project Management Types
 */

export interface Project {
  project_id: number;
  project_name: string;
  client_name?: string;
  project_status: 'Active' | 'On Hold' | 'Completed' | 'Cancelled';
  description?: string;
  approval_status: 'Pending' | 'Approved' | 'Rejected';
  approved_by_manager_id?: number | null;
  approved_at?: string | null;
  approval_note?: string | null;
  can_current_user_approve?: boolean;
  created_at: string;
  manager_name?: string;
  manager_user_id?: number;
  planned_start_date?: string;
  planned_end_date?: string;
  actual_start_date?: string;
  actual_end_date?: string;
}

export interface ProjectTimeline {
  timeline_id: number;
  project_id: number;
  planned_start_date?: string;
  actual_start_date?: string;
  planned_end_date: string;
  actual_end_date?: string;
  version_number: number;
  reason_for_change?: string;
  is_current: boolean;
}

export interface EmployeeProjectAssignment {
  assignment_id: number;
  emp_id: string;
  project_id: number;
  role?: string;
  allocation_pct: number;
  start_date: string;
  end_date?: string;
  status: 'Planned' | 'Active' | 'Ended';
  approval_status?: 'Pending' | 'Approved' | 'Rejected';
  approved_by_manager_id?: number | null;
  approved_at?: string | null;
  approval_note?: string | null;
  approval_requested_by_user_id?: number | null;
  approval_requested_at?: string | null;
  can_current_user_approve?: boolean;
  
  // --- New Billing & Metadata Fields ---
  billing_rate?: number; // Will be null for non-admins due to masking
  billing_start_date?: string;
  billing_end_date?: string;
  is_billable: boolean;
  timesheet_required: boolean;
  client_pm_name?: string;
  billing_project_id?: number;
}

export interface ProjectManagerAssignment {
  id: number;
  project_id: number;
  manager_user_id: number;
  role: string;
  assigned_from: string;
  assigned_to?: string;
  is_active: boolean;
}

// Request Types for Mutations
export type ProjectCreateRequest = Omit<
  Project,
  | 'project_id'
  | 'created_at'
  | 'approval_status'
  | 'approved_by_manager_id'
  | 'approved_at'
  | 'approval_note'
  | 'can_current_user_approve'
  | 'manager_name'
> & {
  manager_user_id?: number;
  planned_start_date?: string;
  planned_end_date?: string;
};
export type ProjectUpdateRequest = Partial<ProjectCreateRequest> & {
  actual_start_date?: string;
  actual_end_date?: string;
  reason_for_change?: string;
};

export interface ProjectApprovalRequest {
  approval_status: 'Approved' | 'Rejected';
  approval_note?: string;
}

export interface AssignmentApprovalRequest {
  approval_status: 'Approved' | 'Rejected';
  approval_note?: string;
}

export interface SkillSearchEmployeeResult {
  emp_id: string;
  full_name: string;
  matched_skills: string[];
  match_type: 'exact' | 'partial' | 'related';
  already_allocated_to_project: boolean;
  allocated_elsewhere: boolean;
  status: 'Available' | 'Already allocated to this project' | 'Allocated elsewhere';
}

export interface SkillSearchResponse {
  query: string;
  limit: number;
  results: SkillSearchEmployeeResult[];
}

export interface SkillRecommendationRequest {
  requested_skills: string[];
  required_count?: number;
  allow_existing_project_assignments?: boolean;
  ai_enabled?: boolean;
}

export interface SkillRecommendationEmployeeResult {
  emp_id: string;
  full_name: string;
  matched_skills: string[];
  related_skills: string[];
  skill_groups: Record<string, string[]>;
  rationale: string;
  relevance_note: string;
  already_allocated_to_project: boolean;
  allocated_elsewhere: boolean;
  status: 'Available' | 'Already allocated to this project' | 'Allocated elsewhere';
}

export interface SkillRecommendationResponse {
  requested_skills: string[];
  required_count: number;
  used_ai_rerank: boolean;
  results: SkillRecommendationEmployeeResult[];
}
export type AssignmentCreateRequest = Omit<
  EmployeeProjectAssignment,
  | 'assignment_id'
  | 'approval_status'
  | 'approved_by_manager_id'
  | 'approved_at'
  | 'approval_note'
  | 'approval_requested_by_user_id'
  | 'approval_requested_at'
  | 'can_current_user_approve'
> & {
  send_for_approval?: boolean;
};

File: backend/db/models/projects.py
from sqlalchemy import Column, Integer, String, Text, Boolean, Date, Numeric, TIMESTAMP, ForeignKey, CheckConstraint, func
from db.base import Base

class Project(Base):
    __tablename__ = "projects"
    project_id = Column(Integer, primary_key=True, autoincrement=True)
    project_name = Column(String(200), nullable=False)
    client_name = Column(String(200))
    project_status = Column(String(30), nullable=False)
    description = Column(Text)
    approval_status = Column(String(20), nullable=False, server_default="Pending")
    approved_by_manager_id = Column(Integer, ForeignKey("users.user_id", ondelete="SET NULL"))
    approved_at = Column(TIMESTAMP)
    approval_note = Column(Text)
    created_at = Column(TIMESTAMP, server_default=func.now(), nullable=False)

    __table_args__ = (
        CheckConstraint("project_status IN ('Active', 'On Hold', 'Completed', 'Cancelled')", name="chk_project_status"),
        CheckConstraint("approval_status IN ('Pending', 'Approved', 'Rejected')", name="chk_project_approval_status"),
    )

class ProjectManager(Base):
    __tablename__ = "project_managers"
    id = Column(Integer, primary_key=True, autoincrement=True)
    project_id = Column(Integer, ForeignKey("projects.project_id", ondelete="CASCADE"), nullable=False)
    manager_user_id = Column(Integer, ForeignKey("users.user_id", ondelete="RESTRICT"), nullable=False)
    role = Column(String(50), nullable=False)
    assigned_from = Column(Date, nullable=False)
    assigned_to = Column(Date)
    is_active = Column(Boolean, server_default="true", nullable=False)

class ProjectTimeline(Base):
    __tablename__ = "project_timelines"
    timeline_id = Column(Integer, primary_key=True, autoincrement=True)
    project_id = Column(Integer, ForeignKey("projects.project_id", ondelete="CASCADE"), nullable=False)
    start_date = Column(Date, nullable=False)
    planned_end_date = Column(Date, nullable=False)
    actual_end_date = Column(Date)
    planned_start_date = Column(Date)
    actual_start_date = Column(Date)
    version_number = Column(Integer, server_default="1", nullable=False)
    reason_for_change = Column(Text)
    is_current = Column(Boolean, server_default="true", nullable=False)

class EmployeeProjectAssignment(Base):
    __tablename__ = "employee_project_assignments"
    assignment_id = Column(Integer, primary_key=True, autoincrement=True)
    emp_id = Column(String(20), ForeignKey("employees.emp_id", ondelete="RESTRICT"), nullable=False)
    project_id = Column(Integer, ForeignKey("projects.project_id", ondelete="CASCADE"), nullable=False)
    role = Column(String(100))
    allocation_pct = Column(Numeric(5, 2), nullable=False)
    start_date = Column(Date, nullable=False)
    end_date = Column(Date)
    status = Column(String(20), nullable=False)
    approval_status = Column(String(20), nullable=False, server_default="Pending")
    approved_by_manager_id = Column(Integer, ForeignKey("users.user_id", ondelete="SET NULL"))
    approved_at = Column(TIMESTAMP)
    approval_note = Column(Text)
    approval_requested_by_user_id = Column(Integer, ForeignKey("users.user_id", ondelete="SET NULL"))
    approval_requested_at = Column(TIMESTAMP)
    
    # --- New Billing Fields ---
    billing_rate = Column(Numeric(10, 2))
    billing_start_date = Column(Date)
    billing_end_date = Column(Date)
    is_billable = Column(Boolean, server_default="true", nullable=False)
    timesheet_required = Column(Boolean, server_default="true", nullable=False)
    client_pm_name = Column(String(200))
    billing_project_id = Column(Integer, ForeignKey("projects.project_id", ondelete="SET NULL"))

    __table_args__ = (
        CheckConstraint("allocation_pct >= 0 AND allocation_pct <= 100", name="chk_allocation_pct"),
        CheckConstraint("status IN ('Planned', 'Active', 'Ended')", name="chk_assignment_status"),
        CheckConstraint(
            "approval_status IN ('Pending', 'Approved', 'Rejected')",
            name="chk_assignment_approval_status",
        ),
    )
